migration_test/migration.test.ts

```diff
@@ -0,0 +1,637 @@
+/*
+ * Copyright 2021, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint-env node */
+'use strict'
+
+import { expect } from 'chai'
+import dotenv from 'dotenv'
+
+import { Wallet } from '@ethersproject/wallet'
+import { parseEther } from '@ethersproject/units'
+import { Provider } from '@ethersproject/abstract-provider'
+
+import { fundL1, fundL2, wait } from '../integration_test/testHelpers'
+import { L2ToL1MessageStatus } from '../src/lib/message/L2ToL1Message'
+import { testSetup } from '../scripts/testSetup'
+import { isNitroL2 } from '../src/lib/utils/migration_types'
+import { BigNumber, Signer } from 'ethers'
+import { ERC20__factory } from '../src/lib/abi/factories/ERC20__factory'
+import { TestWETH9__factory } from '../src/lib/abi/factories/TestWETH9__factory'
+import {
+  EthBridger,
+  getL2Network,
+  Erc20Bridger,
+  IL2ToL1MessageWriter,
+} from '../src'
+dotenv.config()
+
+const waitForNitro = async (l2Provider: Provider, minimum?: number) => {
+  const now = Date.now()
+
+  while (
+    !((await isNitroL2(l2Provider)) && (!minimum || Date.now() - now > minimum))
+  ) {
+    await wait(30000)
+    console.log('Waiting for nitro upgrade')
+  }
+}
+
+abstract class MigrationTest {
+  public name = ''
+  constructor(public readonly testState: TestState) {}
+
+  public abstract initialise(): Promise<void>
+  public abstract finalise(): Promise<void>
+}
```

Comment:
What is the value of using a class here?
I think the code will be cleaner and less dense if we instead declare an object
```ts
interface MigrationTest {
    name: string;
    testState: TestState;
    initialise: () => Promise<void>;
    finalise: () => Promise<void>;
}
```
then MultiTest can receive multiple `MigrationTest` objects